### ISSUE TITLE
Year, Month, and Date extraction from a Date

### ISSUE THREAD

**Issue by s-zeng** [dhall-lang/dhall-lang#1383]
Hello, is it possible to have `Date/year, Date/month, Date/date: Date -> Natural` functions added to the standard? Usecase: well-typed way to have templated directories in the format `.../YYYY/MM/DD/...`

**Comment by kukimik** [comment]
You may have a look at https://github.com/kukimik/dhall-date-format as a possible workaround.

**Comment by s-zeng** [comment]
Wow, did not realize this was doable in pure Dhall. Their `parse` function was exactly what I was looking for
 Thanks!

**Comment by s-zeng** [comment]
Upon further inspections, that package (and Dhall in general) can only extract text components from a date. It would be nice to get integer components

**Comment by kukimik** [comment]
> Upon further inspections, that package (and Dhall in general) can only extract text components from a date.

Yeah, that's why I called it a workaround. I thought it might be sufficient for your use case.

The proposed `Date -> Natural` functions should be easy to include in the standard and implement.  And they would allow to create some utilities (e.g. for date comparison) at the library level.

On the other hand, I'm not sure whether the ability to inspect `Date` values could lead to some bad programming practices. As far as I understand this is the main reason for keeping `Text` opaque.

**Comment by Gabriella439** [comment]
Yeah, I'd support adding functions like these to the standard.  I'd also support the same thing for `Time/{hour,minute,second}`

**Comment by kukimik** [comment]
I can help with the Haskell implementation if anybody (@s-zeng ?) does the standardization work. It's rather easy, but somewhat tedious. Here are some recent examples to follow: https://github.com/dhall-lang/dhall-lang/commit/74233ca6af7485877c6c4fd230ba2dc897730b47, https://github.com/dhall-lang/dhall-lang/pull/1351.

**Comment by kukimik** [comment]
I'm slowly [working on it](https://github.com/kukimik/dhall-lang/tree/datetime-builtins).

What is your opinion on the fractional parts of seconds? Should `Time/second` return seconds truncated to a whole number? Or should we also return the information on the fractional part? If so, what is the preferred way to encode it?

Personally I'd opt for `Time/second : Time -> Natural` returning whole seconds, and forget about the fractional part until somebody actually needs it. Or, if we want to handle it now, I'd rather add another function that would return just the fractional part.

**Comment by kukimik** [comment]
> Maybe a `Time/second : Time -> Double` plus a `Double/truncate : Double -> Integer`?

I'm okay with that. Just note that it makes `Double` a bit less opaque (but `Double`s still can't be manipulated, so I think this is in line with the design choices).

https://docs.dhall-lang.org/discussions/Design-choices.html#arithmetic:

> As floating-point arithmetic is imprecise and prone to surprising results (e.g. loss of associativity/distributivity, loss of accuracy due to cancellation, difficulty testing for equality, etc.), Doubles are opaque values as far as the language is concerned, meaning that a Dhall configuration file can hold them and shuffle them around but cannot manipulate them.

https://hackage.haskell.org/package/dhall-1.42.2/docs/Dhall-Tutorial.html#g:22:

> For Doubles the situation is even more extreme: there are no built-in functions other than Double/show. As far as the language is concerned they are opaque values that can only be shuffled around but not used in any meaningful way until they have been loaded into Haskell.

https://docs.dhall-lang.org/tutorials/Language-Tour.html#numbers:

> On the other hand, Doubles are essentially “opaque”, meaning that you cannot perform any arithmetic with them.

**Comment by kukimik** [comment]
On a second thought, a Dhall implementation is required to [store `Time` with at least nanosecond precision](https://github.com/dhall-lang/dhall-lang/blob/6d4c9d6b8812ed3df1f45aabeca3deaf0615ba7b/standard/dhall.abnf#L575), but I don't think that **maximal** precision is specified anywhere. So we could have an implementation that stores `12:00:59.999999999999999`, but applying `Time/second : Time -> Double` to it would yield `60.0`, which is somewhat alarming.

Also, there is a risk that if in the future some means of constructing `Time` values programmatically (e.g. `Time` arithmetic) or other functions `_ -> Double` are introduced to the language, we may accidentally leak more information about `Double` values than we want.

**Comment by kukimik** [comment]
Related: #1419

If we are switching to unlimited precision, then wouldn't `Time/fractionalSecond` returning a pair of natural numbers (the number after the decimal point and the precision) be the right solution?

I'd rather have `Time/second` return whole seconds, I think it's more useful.

**Comment by winitzki** [comment]
FYI - unlimited precision for fractional seconds has been merged into main branch.
